core/src/main/scala/akka/kafka/internal/SingleSourceLogic.scala

```diff
@@ -78,15 +71,30 @@ import scala.concurrent.{Future, Promise}
     if (!isClosed(shape.out)) {
       complete(shape.out)
     }
-    sourceActor.become {
-      case (_, Terminated(ref)) if ref == consumerActor =>
-        onShutdown()
-        completeStage()
-    }
+    sourceActor.become(shuttingDownReceive)
+    stopConsumerActor()
+  }
+
+  protected def shuttingDownReceive: PartialFunction[(ActorRef, Any), Unit] = {
+    case (_, Terminated(ref)) if ref == consumerActor =>
+      onShutdown()
+      completeStage()
+  }
+
+  protected def stopConsumerActor(): Unit =
     materializer.scheduleOnce(settings.stopTimeout, new Runnable {
       override def run(): Unit =
         consumerActor.tell(KafkaConsumerActor.Internal.Stop, sourceActor.ref)
     })
+
+  protected def partitionAssignedCB(assignedTps: Set[TopicPartition]): Unit = {
```

Comment:
I find the `CB` strange as this is not an `AsyncCallback`, maybe `Signal` or `Handler`?